### ISSUE TITLE
x64 helper call perf optimization opportunities

### ISSUE THREAD

**Issue by AndyAyersMS** [dotnet/coreclr#12940]
Based on comments/code it looks like the jit does not accurately model the kill set of some helpers, eg `CORINFO_HELP_ASSIGN_BYREF` kills very few registers on x64 but the jit models it as a general call plus some extra kills.

The non-windows implementations use `PREPARE_EXTERNAL_VAR` to copy various runtime statics into registers. The macro expand into a RIP-relative memory operand so it would be better to just fold these into the compare instructions directly, eg instead of
```asm
// void JIT_CheckedWriteBarrier(Object** dst, Object* src)
LEAF_ENTRY JIT_CheckedWriteBarrier, _TEXT
        PREPARE_EXTERNAL_VAR g_lowest_address, rax
        cmp     rdi, [rax]
```
one could have
```asm
        cmp     rdi, EXTERNAL_VAR[g_lowest_address]
```
and aside from a more compact and faster-executing helper, this would also free up RAX entirely in several helpers -- which the jit, assuming adopted an accurate model, could then use as a preserved register.

Accurate modelling and fewer kills might lead to proportionally larger wins on Linux because of the larger number of default callee-killed registers.

**Comment by AndyAyersMS** [comment]
Also the helper call properties (set over in utils.cpp) seem to be off for some helpers. Probably worth a top-to-bottom review.

The ones I know about currently:
* `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE` will never return null
* `CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE_MAYBENULL` is missing and so gets worst-case effects. Should be similar to the above but may return null

**Comment by fiigii** [comment]
@AndyAyersMS As I understand, the first step should be folding addressing mode of `PREPARE_EXTERNAL_VAR` into the subsequent instructions, which avoids killing `rax` in Unix implementations. Then we can modify the computing kill set code in LSRA and codgen. Is it right? If yes, could you assign this to me?

**Comment by janvorli** [comment]
@AndyAyersMS, @fiigii we cannot fold the addressing mode as you've suggest.
The `PREPARE_EXTERNAL_VAR` is defined as:
```asm
mov \HelperReg, [rip + C_FUNC(\Name)@GOTPCREL]
```
So the 
```asm
PREPARE_EXTERNAL_VAR g_lowest_address, rax
        cmp     rdi, [rax]
```
is 
```asm
        mov     rax, [rip + g_lowest_address@GOTPCREL]
        cmp     rdi, [rax]
```
As you can see, there is one more indirection than you've expected. The address of the 
g_lowest_address is stored at rip relative address, not its value.

**Comment by AndyAyersMS** [comment]
Hmm, that's too bad. What is it that leads to this extra indirection? Some artifact of PIC?

**Comment by jkotas** [comment]
It is PIC in combination with compiling with default visibility. https://github.com/dotnet/coreclr/issues/2484 talks about fixing it.

It was fixed it in CoreRT and seem to be working great there. You can see that the WriteBarriers in CoreRT do not have the extra indirections: https://github.com/dotnet/corert/blob/master/src/Native/Runtime/amd64/WriteBarriers.S

**Comment by AndyAyersMS** [comment]
@fiigii seems like this is still possible. Still want to give it a try?

I would focus on the runtime side first, the jit part may be somewhat involved. @BruceForstall  has more context on how call kill sets are modeled.

It doesn't look like I can assign this to you myself...maybe somebody else has that ability.

**Comment by fiigii** [comment]
> @fiigii seems like this is still possible. Still want to give it a try?

Of course. Will try to port @jkotas 's visibility change to CoreCLR at first.

**Comment by jkotas** [comment]
> It doesn't look like I can assign this to you myself...maybe somebody else has that ability.

@fiigii I have sent you collaborator invite for CoreCLR so the issue can get assigned to you. Once you accept, make sure to unwatch CoreCLR repo - the git notifications will spam your email otherwise.

**Comment by janvorli** [comment]
@fiigii I have already spent some time trying to make the visibility change for CoreCLR in the past and I have the changes in my private branch. I believe I have that almost complete. You can check the last 4 commits here: https://github.com/janvorli/coreclr/commits/use-hidden-visibility-as-default-2 if you are interested.
I have done some testing with that change and I was not seeing the benefits that @jkotas was referring to, but I didn't spend too much time on that and got distracted by other higher priority work.
If you want to work on top of my changes, I'd be ok with that.

**Comment by fiigii** [comment]
>  Once you accept, make sure to unwatch CoreCLR repo - the git notifications will spam your email otherwise.

Thank you for the reminder 😄 , I did.

**Comment by fiigii** [comment]
> I have already spent some time trying to make the visibility change for CoreCLR in the past and I have the changes in my private branch. I believe I have that almost complete. You can check the last 4 commits here: https://github.com/janvorli/coreclr/commits/use-hidden-visibility-as-default-2 if you are interested.

@janvorli I will investigate into your code. Thank you so much!

**Comment by BruceForstall** [comment]
Well, this GitHub issue is still assigned to you.

**Comment by fiigii** [comment]
@BruceForstall Ok, seems nobody picked it up. I will take a look after I fix 2 known codegen bugs.

**Comment by BruceForstall** [comment]
@fiigii IMO, there are higher priority things that could be worked on, necessary for 3.0. But you can decide.

**Comment by BruceForstall** [comment]
@fiigii I'm going to unassign you. If you still plan to work on it, we can reassign you.
